File: 16-3sum-closest/16-3sum-closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
  if (nums.length <= 3) return nums.reduce((a, b) => a + b)
  nums.sort((a, b) => a - b)
  if (nums[0] >= (target/3 + 1)) return nums[0] + nums[1] + nums[2]
  if (nums[nums.length-1] <= target/3) return nums[nums.length-3] + nums[nums.length-2] + nums[nums.length-1]
  var min = 2000
  for (let i = 0; i < nums.length-2; i++) {
    const a = nums[i]
    const rest = target - a
    var [l, r] = [i+1, nums.length -1]
    while (l < r) {
      const sum = nums[l] + nums[r]
      if (sum < rest) {
        l++
      } else if (sum === rest) {
        return target
      } else {
        r--
      }
      if (Math.abs(sum - rest) < Math.abs(min - target)) {
        min = a + sum
      }
    }
  }
  return min
};

File: 16-3sum-closest/16-3sum-closest.go
func abs(a int) int {
  if a < 0 {
    return -a
  } else {
    return a
  }
}

func threeSumClosest(nums []int, target int) int {
  sort.Ints(nums)
  length := len(nums)
  ans := 3000
  for i := 0; i < length; i++ {
    ai := nums[i]
    l, r := i+1, length-1
    for l < r {
      sum := ai + nums[l] + nums[r]
      if abs(sum - target) < abs(ans - target) {
        ans = sum
      }
      if sum > target {
        r--
      } else {
        l++
      }
    }
  }
  return ans
}